### PR TITLE
[DRAFT] Diffusion: add latent H/W shape bucketing with optional SDPA masking

### PR DESCRIPTION
**Pull request by jy-song-hub**
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

Diffusion serving workloads often receive requests with highly variable image/video resolutions, which translates into many latent H/W shapes and downstream token sequence lengths. This variability makes execution shapes less stable across requests and timesteps, which can reduce reuse opportunities for shape-specialized execution artifacts  (e.g., CUDA Graph capture/reuse, kernel/autotune caches, memory pooling) and increase runtime jitter.                                                   
                                                                                                                                               
This PR introduces a v1 latent-space H/W bucketing mechanism to map requests into a small set of canonical execution shapes. The intent is to improve shape regularity and observability while keeping semantics safe via conservative gating and fallbacks. No performance gain is claimed in this PR.                                                                                                                                  
                                                                                                                                               
**Intuition (systems view)**: Similar to OS “size-class” and caching strategies, we accept small structured overhead (padding/rounding) to reduce shape variance and stabilize the runtime working set. This may make it easier for the runtime to reuse shape-specialized artifacts and can reduce fragmentation/jitter risks, while retaining a safe fallback path when masking is not supported. 

## Modifications

- Add diffusion bucketing CLI flags (v1):                                                                                                    
    •  --enable-diffusion-hw-bucketing                                                                                                         
    •  --diffusion-hw-buckets  (latent  HxW  list)                                                                                             
    •  --diffusion-hw-bucketing-max-waste                                                                                                      
    • File:  python/sglang/multimodal_gen/runtime/server_args.py                                                                               
- Implement core bucketing utilities and types:                                                                                              
    •  BucketMeta  to track original vs bucketed shapes + padding ratio + optional masks                                                       
    • Bucket parsing/selection, H/W padding + unpadding helpers                                                                                
    • Key-padding mask builder for BCTHW patchified tokens (floor-division semantics)                                                          
    • Files:                                                                                                                                   
      •  python/sglang/multimodal_gen/runtime/pipelines_core/types/bucketing.py                                                                
      •  python/sglang/multimodal_gen/runtime/pipelines_core/utils/shape_bucketing.py                                                          
      •  python/sglang/multimodal_gen/runtime/pipelines_core/utils/bucketing_masks.py                                                          
                                                                                                                                               
- Integrate bucketing into denoising (v1 policy + safety):                                                                                   
    • Apply bucketing in  DenoisingStage  for DiT pipelines.                                                                                   
    • If padding exists, attach an attention key-padding mask only when allowed (Torch SDPA + SP=1); otherwise revert to unpadded execution.   
    • Reset per-request bucketing metrics on early-return paths (including the VMOBA fast path) to avoid stale stats when request objects are reused.                                                                                                                                      
    • Cleanup: unpad and clear bucketing metadata at the end of denoising; drop SDPA mask storage to avoid leakage.                            
    • File:  python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py                                                           
- Defensive unpadding at decode time (safety net):                                                                                           
    • If bucketing metadata survives unexpectedly, unpad before VAE decode.                                                                    
    • File:  python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py                                                            
- SDPA backend support for mask consumption:                                                                                                 
    • SDPA consumes canonical  [B, S]  bool mask and broadcasts to the format required by PyTorch SDPA.                                        
    • File:  python/sglang/multimodal_gen/runtime/layers/attention/backends/sdpa.py                                                            
- Plumbing / metrics:                                                                                                                        
    • Track bucketing fields on  Req  ( bucket_id ,  padding_ratio ,  bucket_applied ,  bucket_padded ,  bucket_exact_match ,  bucket_hit ).   
    • Record  unpacked_latent_shape  during latent preparation to infer original THW reliably when packing is used.                            
    • Files:                                                                                                                                   
      •  python/sglang/multimodal_gen/runtime/pipelines_core/schedule_batch.py                                                                 
      •  python/sglang/multimodal_gen/runtime/pipelines_core/stages/latent_preparation.py                                                      
                                                                                                                                               
- Add CPU unit tests:                                                                                                                        
    • Mask generation semantics, token ordering, SDPA mask consumption and validation                                                          
    • Bucketing decision logic (keep vs revert), patch size normalization ( patch_size_t ), and fast-path metric reset behavior                
    • Files:                                                                                                                                   
      •  python/sglang/multimodal_gen/test/test_denoising_bucketing_logic.py                                                                   
      •  python/sglang/multimodal_gen/test/test_sdpa_masked_bucketing_cpu.py   

## Accuracy Tests

This PR affects diffusion execution shapes (via padding) and attention behavior (via masking), so accuracy-oriented validation is included as unit tests:                                                                                                                                  
- pytest -q python/sglang/multimodal_gen/test/test_denoising_bucketing_logic.py -vv                                                         
- pytest -q python/sglang/multimodal_gen/test/test_sdpa_masked_bucketing_cpu.py -vv   

## Benchmarking and Profiling
No speedup claim is made in this PR. Bucketing introduces overhead (padding, larger token counts, mask bookkeeping). Any end-to-end performance impact depends on runtime configuration and how much the system can exploit shape regularity (e.g., reuse/caching). A follow-up benchmarking PR can evaluate latency/throughput under representative serving configurations. 

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 1]
```diff
@@ -129,6 +174,194 @@ def __init__(
...
+
+            if txt_valid is not None:
+                # Canonical convention: True means a valid (non-padding) token.
+                # Casting numeric masks to bool follows PyTorch semantics (nonzero -> True).
+                if txt_valid.dtype != torch.bool:
+                    txt_valid = txt_valid.to(dtype=torch.bool)
+                if txt_valid.device != mask.device:
+                    txt_valid = txt_valid.to(device=mask.device)
+
+                if __debug__:
```
Comment (rainj-me) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:294]: Better not to use __debug__. If the check is solid, just let it to throw the exception.

File: python/sglang/multimodal_gen/runtime/pipelines_core/types/bucketing.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+@dataclass
+class BucketMeta:
+    """Metadata for a bucketed diffusion request.
+
+    v1 primarily targets DiT-style HW bucketing, but the same container is also
+    used by utilities that operate on packed-token layouts (e.g. "BSD").
+    """
+
+    enabled: bool
+    layout: str  # e.g. "BCTHW", "BCHW", "BSD"
```
Comment (rainj-me) [inline python/sglang/multimodal_gen/runtime/pipelines_core/types/bucketing.py:15]: maybe enum.

File: python/sglang/multimodal_gen/runtime/pipelines_core/utils/bucketing_masks.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import torch
+
+from sglang.multimodal_gen.runtime.platforms import AttentionBackendEnum
+
+
+def build_key_padding_mask_bcthw(
+    B: int,
+    exec_thw: tuple[int, int, int],
```
Comment (rainj-me) [inline python/sglang/multimodal_gen/runtime/pipelines_core/utils/bucketing_masks.py:8]: rely on Tupple type from typing.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 4]
```diff
@@ -494,6 +727,216 @@ def _postprocess_latents_for_ti2v(self, z, reserved_frames_masks, batch):
...
+            )
+            bucketing_log = logger.info if bucketing_bench_logs else logger.debug
+
+            orig_thw = get_orig_thw(bucket_meta)
+
+            if bucket_meta.enabled and bucket_meta.padding_ratio > 0.0:
+                if self._should_use_bucketing_with_masks():
+                    assert (
+                        get_sp_world_size() == 1
+                    ), "Masked bucketing is not supported with sequence parallelism."
```
Comment (rainj-me) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:824]: Is the assert required since the if condition already limit sp_world_size == 1

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 5]
```diff
@@ -494,6 +727,216 @@ def _postprocess_latents_for_ti2v(self, z, reserved_frames_masks, batch):
...
+                batch.exec_latent_shape = (
+                    int(bucket_meta.orig_t),
+                    int(bucket_meta.bucket_h),
+                    int(bucket_meta.bucket_w),
+                )
+
+            bucketing_bench_logs = get_bool_env_var(
+                "SGLANG_DIFFUSION_BUCKETING_BENCH_LOGS", "false"
+            )
+            bucketing_log = logger.info if bucketing_bench_logs else logger.debug
```
Comment (rainj-me) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:816]: This code is weird.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 6]
```diff
@@ -494,6 +727,216 @@ def _postprocess_latents_for_ti2v(self, z, reserved_frames_masks, batch):
...
+                if isinstance(reqs, (list, tuple)):
+                    for req in reqs:
+                        req.bucket_id = bucket_id
+                        req.padding_ratio = padding_ratio
+
+                        req.bucket_applied = bucket_applied
+                        req.bucket_padded = bucket_padded
+                        req.bucket_exact_match = bucket_exact_match
+
+                        req.bucket_hit = bucket_padded
```
Comment (rainj-me) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:892]: Please put the bucket info to a sperate object instead of spreading to req obj.
